internal/credential/vault/repository_credential_library.go

```diff
@@ -0,0 +1,74 @@
+package vault
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/oplog"
+)
+
+// CreateCredentialLibrary inserts l into the repository and returns a new
+// CredentialLibrary containing the credential library's PublicId. l is not
+// changed. l must contain a valid StoreId. l must not contain a PublicId.
+// The PublicId is generated and assigned by this method.
+//
+// Both l.Name and l.Description are optional. If l.Name is set, it must be
+// unique within l.StoreId.
+//
+// Both l.CreateTime and l.UpdateTime are ignored.
+func (r *Repository) CreateCredentialLibrary(ctx context.Context, scopeId string, l *CredentialLibrary, _ ...Option) (*CredentialLibrary, error) {
+	const op = "vault.(Repository).CreateCredentialLibrary"
+	if l == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "nil CredentialLibrary")
+	}
+	if l.CredentialLibrary == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "nil embedded l")
+	}
+	if l.StoreId == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "no store id")
+	}
+	if l.VaultPath == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "no vault path")
+	}
+	if l.PublicId != "" {
+		return nil, errors.New(errors.InvalidParameter, op, "public id not empty")
+	}
+	if scopeId == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "no scope id")
+	}
+	l = l.clone()
```

Comment:
You are calling clone here and then again on line 58 one of them could be dropped

Reply:
We do [this](https://github.com/hashicorp/boundary/blob/93d81157758398255baf9715a38fa0b381d975cb/internal/host/static/repository_host_catalog.go#L37) in [other](https://github.com/hashicorp/boundary/blob/93d81157758398255baf9715a38fa0b381d975cb/internal/host/static/repository_host_catalog.go#L71) places too. If I remember correctly, the first clone prevents changes to the object passed in and the second clone is needed for the `w.Create` method.